sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/Plugin.java

```diff
@@ -83,46 +86,50 @@ public Plugin withConfig(PluginConfig pluginConfig) {
    * validating connection to the CDAP sink/source and performing initial tuning.
    */
   public void prepareRun() {
-    PluginConfig pluginConfig = getPluginConfig();
-    checkStateNotNull(pluginConfig, "PluginConfig should be not null!");
-    if (cdapPluginObj == null) {
-      try {
-        Constructor<?> constructor =
-            getPluginClass().getDeclaredConstructor(pluginConfig.getClass());
-        constructor.setAccessible(true);
-        cdapPluginObj = (SubmitterLifecycle) constructor.newInstance(pluginConfig);
-      } catch (Exception e) {
-        LOG.error("Can not instantiate CDAP plugin class", e);
-        throw new IllegalStateException("Can not call prepareRun");
-      }
-    }
-    try {
-      cdapPluginObj.prepareRun(getContext());
-      if (getPluginType().equals(PluginConstants.PluginType.SOURCE)) {
-        for (Map.Entry<String, String> entry :
-            getContext().getInputFormatProvider().getInputFormatConfiguration().entrySet()) {
-          getHadoopConfiguration().set(entry.getKey(), entry.getValue());
+    if (!isUnbounded()) {
+      PluginConfig pluginConfig = getPluginConfig();
+      checkStateNotNull(pluginConfig, "PluginConfig should be not null!");
+      if (cdapPluginObj == null) {
+        try {
+          Constructor<?> constructor =
+              getPluginClass().getDeclaredConstructor(pluginConfig.getClass());
+          constructor.setAccessible(true);
+          cdapPluginObj = (SubmitterLifecycle) constructor.newInstance(pluginConfig);
+        } catch (Exception e) {
+          LOG.error("Can not instantiate CDAP plugin class", e);
+          throw new IllegalStateException("Can not call prepareRun");
         }
-      } else {
-        for (Map.Entry<String, String> entry :
-            getContext().getOutputFormatProvider().getOutputFormatConfiguration().entrySet()) {
-          getHadoopConfiguration().set(entry.getKey(), entry.getValue());
+      }
+      try {
+        cdapPluginObj.prepareRun(getContext());
+        if (getPluginType().equals(PluginConstants.PluginType.SOURCE)) {
+          for (Map.Entry<String, String> entry :
```

Comment:
Personally, I think this is unclear what you are doing here and why. Plus these for loops look almost identical, except for the collection part.

Reply:
For `Source` plugins we are copying configuration from `InputFormatProvider`, for `Sink` plugins from `OutputFormatProvider`. 
Do you suggest moving it to a separate method?